CONTRIBUTING.md

```diff
@@ -0,0 +1,110 @@
+# Rodio Development Guide
+
+## Quick Start
+
+1. Install [Rust](https://www.rust-lang.org/tools/install)
+2. Clone the repository: `git clone https://github.com/RustAudio/rodio.git`
+3. Navigate to the project: `cd rodio`
+4. Build the project: `cargo build`
+
+## Project Structure
+
+src/:
+- `source/`: Audio source implementations
+- `decoder/`: Audio format decoders
+- `sink/`: Audio playback and mixing
+- `dynamic_mixer/`: Real-time audio mixing
+- `spatial/`: Spatial audio capabilities
+
+## Coding Guidelines
+
+- Follow [Rust API Guidelines](https://rust-lang.github.io/api-guidelines/)
+- Use `rustfmt` for formatting
+- Implement `Source` trait for new audio sources
+- Use `Sink` for playback management
+
+## Common Tasks
+
+### Adding a New Audio Source or Effect
+
+1. Create a new file in `src/source/`
+2. Implement the `Source` trait to define how audio samples are generated or modified
+3. Consider implementing sources like oscillators, noise generators, or effects like amplification, filtering, or distortion
+4. Optimize for performance, especially for real-time processing in effects
+5. Add unit tests/benchmarks where applicable to ensure proper functionality and performance
+
+### Implementing a New Decoder
+
+1. Add new module in `src/decoder/`
+2. Implement necessary traits (e.g., `Decoder`) to handle specific audio formats
+3. Focus on efficiently parsing audio file headers and decoding compressed audio data
+4. Update `src/decoder/mod.rs` to integrate the new decoder
+
+## Testing
+
+- Write unit tests for each new function (if applicable)
+- Add integration tests for end-to-end scenarios
```

Comment:
"If possible add an integration tests. Note: normally these do not need to produce any sound (in other words use the sink/stream). If your adding a source check if the samples have changed, are non zero etc. Unfortunately most things in audio can not easily be checked automatically. For those adding an example that does produce sound and listening to it works.

Reply:
This what you want me to copy and paste in verbatim or paraphrased?

Reply:
you could, but do feel free to touch it up.